logp/typedloggercore.go

```diff
@@ -0,0 +1,84 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package logp
+
+import (
+	"fmt"
+
+	"go.uber.org/zap/zapcore"
+)
+
+// typedLoggerCore takes two cores and directs logs entries to one of them
+// with the value of the field defined by the pair `key` and `value`
+//
+// If `entry[key] == value` the typedCore is used, otherwise the
+// defaultCore  is used.
+// WARNING: The level of both cores must always be the same!
+// typedLoggerCore will only use the defaultCore level to decide
+// whether to log an entry or not
+type typedLoggerCore struct {
+	typedCore   zapcore.Core
+	defaultCore zapcore.Core
+	value       string
+	key         string
+}
+
+func (t *typedLoggerCore) Enabled(l zapcore.Level) bool {
+	return t.defaultCore.Enabled(l)
+}
+
+func (t *typedLoggerCore) With(fields []zapcore.Field) zapcore.Core {
+	newCore := typedLoggerCore{
+		defaultCore: t.defaultCore.With(fields),
+		typedCore:   t.typedCore.With(fields),
+		key:         t.key,
+		value:       t.value,
+	}
+	return &newCore
+}
+
+func (t *typedLoggerCore) Check(e zapcore.Entry, ce *zapcore.CheckedEntry) *zapcore.CheckedEntry {
+	if t.defaultCore.Enabled(e.Level) {
+		return ce.AddCore(e, t)
+	}
+
+	return ce
+}
+
+func (t *typedLoggerCore) Sync() error {
+	defaultErr := t.defaultCore.Sync()
+	typedErr := t.typedCore.Sync()
+
+	if defaultErr != nil || typedErr != nil {
+		return fmt.Errorf("error syncing logger. DefaultCore: '%w', typedCore: '%w'", defaultErr, typedErr)
+	}
+
+	return nil
+}
+
+func (t *typedLoggerCore) Write(e zapcore.Entry, fields []zapcore.Field) error {
+	for _, f := range fields {
+		if f.Key == t.key {
+			if f.String == t.value {
+				return t.typedCore.Write(e, fields)
+			}
```

Comment:
> can we have multiple (f.Key == t.key) hits here?

No, `f.Key` need to be unique, otherwise they will end up as duplicated in the final JSON. While the code will accept it, there will be issues down the road.

Another reason to do not duplicate is is that `t.key/t.vlue` indicates the log type, so there can be only one set. A log entry cannot be `default` and `event` at the same time.

Reply:
so we can break a loop once we hit f.Key==t.key and inner condition is not met

Reply:
Done in 1dca25a.

Thanks for spotting that!